### ISSUE TITLE
NVDA is skipping focus for menu items when using up and down cursor

### ISSUE THREAD

**Issue by smokymcbear** [nvaccess/nvda#7723]
### Steps to reproduce:
 Firefox 56.0.2 (32-bit)
 Windows 10 64-bit

 `<Button onKeydown={(event) => {console.log("onKeydown works!")}} />`
 The above console.log isn't outputting at all when skipping through the menu items. The Screen reader also seems to read the menu items it has skipped too. 
 Navigating the menu with cursors at a fast speed will force the screen reader to not skip focus however navigating at a slow speed will cause the screen reader to skip focus.
 Turning off NVDA will cause the menu items to be navigated through properly in Firefox.

![nvdaerror](https://user-images.githubusercontent.com/19962586/32355166-58a5ba40-bfea-11e7-9741-9c4421eaabd4.gif)

### Expected behavior:
Menu items should be read in order and should be able to navigate along the menu without skipping

### Actual behavior:
Menu items are being skipped when navigating along the menu

### System configuration:
NVDA version: 2017.3

NVDA Installed or portable: NVDA installed

Other information:
Running in a VM

Windows version:
Windows 10 64-bit

Name and version of other software in use when reproducing the issue:
Javascript ES6 with React.js

### Other questions:

Does the issue still occur after restarting your PC? 
Yes

Have you tried any other versions of NVDA?
No
> Please list them and the result

**Comment by feerrenrut** [comment]
@smokymcbear Please note, many of the NVDA developers are blind and are unable to view the GIF in the description. Furthermore, we can only guess as to how the menu in the GIF is implemented.

**Could you please provide a minimal example, with instructions for how you interact with it?**
For example, please describe the keypresses used to move through the menu (arrow keys or tab).

**Comment by smokymcbear** [comment]
Hi @feerrenrut , here's a minimal example of the code. I used arrow keys UP and DOWN to navigate through the menu. Here's an [interactive example](https://codesandbox.io/s/pyy5mx44qx). Please let me know if there's anymore information I can provide to diagnose this situation. Thanks!

```
import React, { Component } from "react";
import { render } from 'react-dom';
import PropTypes from "prop-types";


const keyCode = {
  UP: 38,
  DOWN: 40,
  TAB: 9
};

class ExampleMenu extends Component {
  constructor(props: Object): void {
    super(props);

    this._renderMenu = this._renderMenu.bind(this);
  }

  _focusButton(index) {
    const focusButton = document.querySelector(`#menuItem-${index}`);

    if (focusButton) {
      focusButton.focus();
    }
  }

  _onButtonKeyDown(e, index) {
    switch (e.keyCode) {
      case keyCode.DOWN: {
        if (index !== this.numberOfMenuItems - 1) {
          this._focusButton(index + 1);
        } else {
          this._focusButton(0)
        }
      }
      case keyCode.UP: {
        if (index === 0) {
          this._focusButton(this.numberOfMenuItems - 1)
        } else {
          this._focusButton(index - 1)
        }
      }
      default:
        break;
    }
  }

  _renderMenu(menuItems: Array<Object>): Array<ReactElement> {
    this.numberOfMenuItems = menuItems.length;

    return menuItems.map((menuItem, index) => {
      return (
        <button
         style={{display:"block"}}
          key={index}
          id={`menuItem-${index}`}
          onKeyDown={event => {
            this._onButtonKeyDown(event, index);
          }}
        >
          {menuItem.name}
        </button>
      );
    });
  }

  render(): ReactElement {
    const { menuItems } = this.props;

    return (
      <div>
        {this._renderMenu(menuItems)}
      </div>
    );
  }
}

ExampleMenu.propTypes = {
  menuItems: PropTypes.array.isRequired
}

export default ExampleMenu;
```

**Comment by smokymcbear** [comment]
Ok so within the interactive example https://codesandbox.io/s/pyy5mx44qx , the example code isn't exactly working as intended on either chrome or firefox. However, when I turn on NVDA, the code is working as intended which is very strange. Perhaps, NVDA is registering different keycode events?

**Comment by derekriemer** [comment]
As far as I know, NVDA isn't messing with the key codes. Are you sure your NVDA is in focus mode?

**Comment by smokymcbear** [comment]
As far as I know, I'm using the default settings for NVDA since I have not touched any settings since downloading it. Does the default settings include being defaulted to focus mode? 

I'm currently looking around the settings/preferences and I don't happen to see a focus mode. I only see a browse mode. @derekriemer May you please direct me to activate focus mode? 

Sorry but the example I provided before didn't have the `console.log()` within the switch case. Please try https://codesandbox.io/s/r54j2mqrl4 again. Are you seeing anything in the console when NVDA is on? It's only logging into the console when I do **not** have NVDA on, however when NVDA is on, nothing is being logged into the console which leads me to believe the keycode is not registering.

**Comment by derekriemer** [comment]
you should read section [6 of the users guide](https://www.nvaccess.org/files/nvda/documentation/userGuide.html).

**Comment by smokymcbear** [comment]
I see, thank you for the guide and apologies for my ignorance. Yes, @derekriemer NVDA is on focus mode.

**Comment by derekriemer** [comment]
NVDA is in focus mode, and it isn't working as expected? When I tried testing this in focus mode, it worked fine.

**Comment by smokymcbear** [comment]
While in NVDA focus mode, the `console.log()` is not registering but for some reason, I'm able to correctly arrow key up and down accordingly.

While **not** in NVDA focus mode, the `console.log()` is registering and the order only goes up regardless of whether I use arrow key up or down(sorry for the poor code example but it still shows the problem I'm having with NVDA).

**Comment by derekriemer** [comment]
Did you change the code example to emit the console.log? Sorry, I wasn't aware it had changed. The general idea is that your key press won't fire while in browse mode, because the user is dealing with a virtual document.

**Comment by smokymcbear** [comment]
@derekriemer Yep, I just added a `console.log()` within the switch case statement for both `keycode.UP` and `keycode.DOWN`. 

To be more specific with this issue, when I said previously, "While **not** in NVDA focus mode", I mean that while I am not running the NVDA screen reader program at all, the `console.log()` is being registered and while I am running NVDA in focus mode, the `console.log()` is not being registered. Does this happen for you also? @derekriemer 

I hope this helps in diagnosing this issue.

**Comment by feerrenrut** [comment]
Firstly there is a bug in https://codesandbox.io/s/r54j2mqrl4
```
case keyCode.DOWN: {
        console.log("DOWN", index);
        if (index !== this.numberOfMenuItems - 1) {
          this._focusButton(index + 1);
        } else {
          this._focusButton(0)
        }
      }
```
the `break` is missing. You are falling through to the `keyCode.UP` case. Meaning, pressing the down key first runs the keydown code, and then the key up code.

For me this code has the same behaviour without NVDA and with NVDA while in focus mode.

If you are new to NVDA:
- I would suggest turning off the audiocons for browse and focus mode changes. 
Do this by disabling the option for "audio indication of focus and browse mode" in the NVDA menu, preferences, browse mode dialog.
- Using `NVDA+space` to ensure you are in the right mode before testing
- Using the speechviewer to be able to read through the history of things announced by NVDA, enable this in the NVDA menu, tools, speech viewer.

**Comment by smokymcbear** [comment]
Hi @feerrenrut , I believe I found the cause of this issue. 

I am currently developing on a mac while running a VM of windows 10. Because I was using a VM to run the windows 10 environment, for some reason I was not able to capture the caps lock + space key press in the VM, thus not enabling focus mode. 

Just today, I have gained access to a tangible windows 8 device to which I've tested the code example and my application where I'm experiencing this issue. Focus mode is registering the keycodes correctly, while browse mode is not registering the keycodes as expected. I guess you just can't simulate the real thing sometimes. 

Thank you for all your help and especially your time + effort @feerrenrut @derekriemer.  I will be closing this issue since this issue is due to my environment and not the NVDA software itself.
